implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs

```diff
@@ -0,0 +1,314 @@
+use alloc::sync::Arc;
+use core::time::Duration;
+use ockam_core::compat::{boxed::Box, vec::Vec};
+use ockam_core::{
+    Address, AllowAll, Any, Decodable, DenyAll, LocalOnwardOnly, LocalSourceOnly, Mailbox,
+    Mailboxes, OutgoingAccessControl, Route, Routed,
+};
+use ockam_core::{AllowOnwardAddress, Result, Worker};
+use ockam_node::callback::CallbackSender;
+use ockam_node::{Context, WorkerBuilder};
+use tracing::{debug, info, trace};
+
+/// This struct implements a Worker receiving and sending messages
+/// on one side of the secure channel creation as specified with its role: INITIATOR or REPSONDER
+pub(crate) struct HandshakeWorker {
+    secure_channels: Arc<SecureChannels>,
+    callback_sender: CallbackSender<()>,
+    state_machine: Box<dyn StateMachine>,
+    identifier: IdentityIdentifier,
+    addresses: Addresses,
+    role: Role,
+    remote_route: Option<Route>,
+    decryptor: Option<DecryptorWorker>,
+}
+
+#[ockam_core::worker]
+impl Worker for HandshakeWorker {
+    type Message = Any;
+    type Context = Context;
+
+    /// Initialize the state machine with an `Initialize` event
+    /// Depending on the state machine role there might be a message to send to the other party
+    async fn initialize(&mut self, context: &mut Self::Context) -> Result<()> {
+        match self.state_machine.on_event(Initialize).await? {
+            SendMessage(message) => {
+                info!(
+                    "remote route {:?}, decryptor remote {:?}",
+                    self.remote_route.clone(),
+                    self.addresses.decryptor_remote.clone()
+                );
+                context
+                    .send_from_address(
+                        self.remote_route(),
+                        message,
+                        self.addresses.decryptor_remote.clone(),
+                    )
+                    .await
+            }
+            Action::NoAction => Ok(()),
+        }
+    }
+
+    /// Handle a message coming from the other party
+    /// If the handshake has been fully performed then we can delegate this message to the
+    /// secure channel Decryptor.
+    /// Otherwise we unpack the message payload and send it to the state machine to trigger
+    /// a transition
+    async fn handle_message(
+        &mut self,
+        context: &mut Self::Context,
+        message: Routed<Self::Message>,
+    ) -> Result<()> {
+        if let Some(decryptor) = self.decryptor.as_mut() {
+            // since we cannot replace this worker with the decryptor worker we act as a relay instead
+            return decryptor.handle_message(context, message).await;
+        };
+
+        // we only set it once to avoid redirects attack
+        self.remote_route = Some(message.return_route());
+        if let SendMessage(message) = self
+            .state_machine
+            .on_event(ReceivedMessage(Vec::<u8>::decode(
+                &message.into_transport_message().payload,
+            )?))
+            .await?
```

Comment:
How should we handle error here? Should we shutdown the worker?

Reply:
We should but we don't have the machinery to do it properly. This is not something that can be tackled properly in this PR unfortunately.